### PR TITLE
End-to-end tests for post-training API

### PR DESCRIPTION
**Pull request by AlexKoff88**
### Changes

Added e2e test that takes models from Timm package and run quantize() API for PyTorch, ONNX and OpenVINO representations.
Resulted models are converted to OV IR and benchmarked. Final results are reported in .csv file.

### CODE REVIEW

File: tests/post_training/conftest.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import pytest
+import numpy as np
+from pathlib import Path
+
+def pytest_addoption(parser):
+    parser.addoption("--data", action="store", default="/mnt/datasets/imagenet/val")
+    parser.addoption("--output", action="store", default="./tmp/")
```
Comment (alexsu52) [inline tests/post_training/conftest.py:7]: Please, set `default=None`

**Review by alexsu52 — APPROVED** [review]
Cool staff!

File: tests/post_training/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--extra-index-url https://download.pytorch.org/whl/cpu
+torch
+--extra-index-url https://download.pytorch.org/whl/cpu
+torchvision
+onnx
+onnxruntime
+pytest
+openvino>=2022.1
+openvino-dev>=2022.1
```
Comment (andrey-churkin) [inline tests/post_training/requirements.txt:9]:
```suggestion
openvino-dev>=2022.1
```

Comment (andrey-churkin) [reply tests/post_training/requirements.txt:9]: > Installing OpenVINO Development Tools will also install OpenVINO Runtime as a dependency, so you don’t need to install OpenVINO Runtime separately.

https://docs.openvino.ai/latest/openvino_docs_install_guides_install_dev_tools.html#for-python-developers

File: tests/post_training/test_quantize_conformance.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    export_to_ir(onnx_path, output_path, model_name)
+
+    # Benchmark performance
+    performance = benchmark_performance(ov_path, model_name)
+    # Validate accuracy
+    accuracy = validate_accuracy(ov_path, dataloader)
+    return performance, accuracy
+
+
+def benchmark_ov_model(model, dataloader, model_name, output_path):
```
Comment (vshampor) [inline tests/post_training/test_quantize_conformance.py:180]: All of the "benchmark" functions measure both accuracy and performance. I don't think that this is flexible. This could be separated by test case parametrization.

Furthermore, without [special host preparation](https://easyperf.net/blog/2019/08/02/Perf-measurement-environment-on-Linux) both in and out of this test code the performance measurements might not be consistent. It might not be possible to compare the perf results across runs and even across the frameworks in the same run. Complete instructions on how to setup the host state for correct performance measurements, or a reference to a script should be included in the README.

File: tests/post_training/conftest.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+@pytest.hookimpl(tryfirst=True, hookwrapper=True)
+def pytest_runtest_makereport(item, call):
+    outcome = yield
+    result = outcome.get_result()
+
+    if result.when == 'call':
+        table = item.config.test_results
+        header = ["Model name", "FP32 top-1", "Torch INT8 top-1", "ONNX INT8 top-1", "OV INT8 top-1",
+            "FP32 FPS", "Torch INT8 FPS", "ONNX INT8 FPS", "OV INT8 FPS"]
```
Comment (vshampor) [inline tests/post_training/conftest.py:20]: Looks fragile. I think you should rather set up a `BenchmarkResult` dataclass with fields such as `model`, `framework`, `result_type` (accuracy or performance) and `value`; you can also add the `succeeded` and `error_msg` fields for error handling if you want to see it in the report. You can then group the results and make a table in this function. This will remove the need for awkward `result.append([model_name, error, "-", "-", "-", "-", "-", "-", "-"])` stuff where one needs to put the exactly correct number of `"-"` elements in the other file to always correspond to the header length in this file.

File: tests/post_training/test_quantize_conformance.py [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+
+
+def export_to_ir(model_path, save_path, model_name):
+    command_line = f"mo -m {model_path} -o {save_path} -n {model_name}"
+    os.popen(command_line).read()
+
+
+def run_benchmark(model_path):
+    command_line = f"benchmark_app -m {model_path} -d CPU -niter 300"
+    cmd_output = os.popen(command_line).read()
```
Comment (vshampor) [inline tests/post_training/test_quantize_conformance.py:88]: We have `tests.common.command.Command` for that. Same for all other spots where a command has to be executed in the shell.

File: tests/post_training/test_quantize_conformance.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        model.eval().cpu()
+        transform = get_model_transform(model)
+
+        batch_one_dataloader = get_torch_dataloader(data, transform, batch_size=1)
+        # benchmark original models (once)
+        orig_perf, orig_acc = benchmark_torch_model(
+            model, batch_one_dataloader, model_name, output_folder
+        )
+
+        val_dataloader = get_torch_dataloader(data, transform, batch_size=128)
```
Comment (vshampor) [inline tests/post_training/test_quantize_conformance.py:229]: What's the significance of the batch size 128? Should we make this a parameter of the pytest run? Should we align it across the frameworks in this test?

**Review by vshampor — CHANGES_REQUESTED** [review]

File: tests/post_training/README.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# Post-training Quantization Confirmance Suite
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/README.md:1]: See the description below.

[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/README.md:1]: There is no intend to extend it to non PTQ scenarios. It takes a lot of time, btw even in post-training.

File: tests/post_training/README.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+...
+```
+
+## Usage
+Once the evnironment is installed use the following command to run the test:
+```
+NUM_VAL_THREADS=8 pytest --data=<path_to_imagenet_val_folder> --output=./tmp tests/post_training/test_quantize_conformance.py
+```
+
+`NUM_VAL_THREADS` environment variable controls the number of parallel streams when validating the model.
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/README.md:30]: This was done intentionally since it is an environment attribute not the test.

File: tests/post_training/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (vshampor) [reply tests/post_training/README.md:1]: I don't see how, by the analysis of the results of these tests, a decision can be made whether NNCF's post-training functionality conforms or does not conform to something.

File: tests/post_training/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (vshampor) [reply tests/post_training/README.md:30]: It is a parameter influencing the test that you seem to want to expose to the user of the test. This should be done via the conventional way of defining a pytest option, since we control the test ourselves. Relying on environment variables is essentially using global variables. You don't use an environment variable for determining the output directory, do you?

File: tests/post_training/conftest.py [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import pytest
+import numpy as np
+from pathlib import Path
+
+def pytest_addoption(parser):
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/conftest.py:5]: Won't fix as minor.

File: tests/post_training/conftest.py (hunk 4 above)
Comment (AlexKoff88) [reply tests/post_training/conftest.py:20]: Won't fix as minor.

File: tests/post_training/model_scope.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
+from typing import List
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/model_scope.py:1]: Won't fix as minor.

File: tests/post_training/test_quantize_conformance.py (hunk 5 above)
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:88]: There is no type hint, help, need to understand how it works but valid. Reworked.

File: tests/post_training/test_quantize_conformance.py (hunk 3 above)
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:180]: Won't fix as a minor.

File: tests/post_training/test_quantize_conformance.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+def benchmark_ov_model(model, dataloader, model_name, output_path):
+    # Dump model
+    ov_path = Path(output_path) / (model_name + ".xml")
+    ov.serialize(model, str(ov_path))
+
+    # Benchmark performance
+    performance = benchmark_performance(ov_path, model_name)
+    # Validate accuracy
+    accuracy = validate_accuracy(ov_path, dataloader)
+    return performance, accuracy
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:189]: Won't fix as a minor.

File: tests/post_training/test_quantize_conformance.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    return pytestconfig.getoption("output")
+
+
+@pytest.fixture(scope="session")
+def result(pytestconfig):
+    return pytestconfig.test_results
+
+
+@pytest.mark.parametrize("model_args", get_validation_scope())
+def test_ptq_timm(data, output, result, model_args): # pylint: disable=W0703
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:208]: Won't fix as a minor.

File: tests/post_training/test_quantize_conformance.py (hunk 6 above)
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:229]: It's some reasonable default.

File: tests/post_training/test_quantize_conformance.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+
+
+@pytest.fixture(scope="session")
+def output(pytestconfig):
+    return pytestconfig.getoption("output")
+
+
+@pytest.fixture(scope="session")
+def result(pytestconfig):
+    return pytestconfig.test_results
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:204]: Won't fix as a minor.

File: tests/post_training/test_quantize_conformance.py [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+
+
+@pytest.mark.parametrize("model_args", get_validation_scope())
+def test_ptq_timm(data, output, result, model_args): # pylint: disable=W0703
+    torch.multiprocessing.set_sharing_strategy(
+        "file_system"
+    )  # W/A to avoid RuntimeError
```
[the remark this answers is not in the document]
Comment (AlexKoff88) [reply tests/post_training/test_quantize_conformance.py:211]: Will take it into account an use in the future.

**Comment by AlexKoff88** [comment]
run build

**Review by vshampor — COMMENTED** [review]
I do not approve of the code quality of this PR. In the state the test is now I do not see its meaningful usage and evolution past manual runs.
